patches/api/0421-Add-source-block-to-the-BlockGrowEvent.patch

```diff
@@ -0,0 +1,100 @@
+     public HandlerList getHandlers() {
+diff --git a/src/main/java/org/bukkit/event/block/BlockGrowEvent.java b/src/main/java/org/bukkit/event/block/BlockGrowEvent.java
+index 73d5375d2f67ee00204d80308828cd9f2b2f17e8..13d03f3d59ef147b48a3da3e5287960199915f7a 100644
+--- a/src/main/java/org/bukkit/event/block/BlockGrowEvent.java
++++ b/src/main/java/org/bukkit/event/block/BlockGrowEvent.java
+@@ -24,12 +24,21 @@ import org.jetbrains.annotations.NotNull;
+ public class BlockGrowEvent extends BlockEvent implements Cancellable {
+     private static final HandlerList handlers = new HandlerList();
+     private final BlockState newState;
++    private final Block source; // Paper
+     private boolean cancelled = false;
+ 
++    @Deprecated // Paper
+     public BlockGrowEvent(@NotNull final Block block, @NotNull final BlockState newState) {
++        this(block, null, newState); // Paper
++    }
++
++    // Paper start
++    public BlockGrowEvent(@NotNull final Block block, @org.jetbrains.annotations.Nullable final Block source, @NotNull final BlockState newState) { // Paper
+         super(block);
++        this.source = source;
+         this.newState = newState;
+     }
++    // Paper end
+ 
+     /**
+      * Gets the state of the block where it will form or spread to.
+@@ -41,6 +50,20 @@ public class BlockGrowEvent extends BlockEvent implements Cancellable {
+         return newState;
+     }
+ 
++    // Paper start
++    /**
++     * Gets the source block involved in this event.
++     * The block can be null when the block grow without
++     * creating a new block but instead modifying its properties.
++     *
++     * @return the source block involved in this event.
++     */
++    @org.jetbrains.annotations.Nullable
++    public Block getSource() {
++        return this.source;
++    }
++    // Paper end
++
+     @Override
+     public boolean isCancelled() {
+         return cancelled;
+diff --git a/src/main/java/org/bukkit/event/block/BlockSpreadEvent.java b/src/main/java/org/bukkit/event/block/BlockSpreadEvent.java
+index e9239caec9d20985479f090bff15c7f2ec90c8cd..0050fcd871244b14deb5da7e4d2454f0cd40459b 100644
+--- a/src/main/java/org/bukkit/event/block/BlockSpreadEvent.java
++++ b/src/main/java/org/bukkit/event/block/BlockSpreadEvent.java
+@@ -23,11 +23,8 @@ import org.jetbrains.annotations.NotNull;
+  */
+ public class BlockSpreadEvent extends BlockFormEvent {
+     private static final HandlerList handlers = new HandlerList();
+-    private final Block source;
+-
+     public BlockSpreadEvent(@NotNull final Block block, @NotNull final Block source, @NotNull final BlockState newState) {
+-        super(block, newState);
+-        this.source = source;
++        super(block, source, newState); // Paper
+     }
+ 
+     /**
+@@ -37,7 +34,7 @@ public class BlockSpreadEvent extends BlockFormEvent {
+      */
+     @NotNull
+     public Block getSource() {
+-        return source;
++        return super.getSource(); // Paper
```

Comment:
See what I said above, I think you can leave this, and just override getSource() to return the field in this event instead of passing it up the inheritance chain.